styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -208,7 +168,120 @@ private void ensureSecrets(WorkflowInstance workflowInstance, RunSpec runSpec) {
         LOG.info("[AUDIT] Workflow {} refers to secret {}",
                  workflowInstance.workflowId(), specSecret.name());
       }
+    });
+  }
+
+  private void ensureServiceAccountKeySecret(WorkflowInstance workflowInstance, RunSpec runSpec) {
+    runSpec.serviceAccount().ifPresent(serviceAccount -> {
+      final String secretName = buildSecretName(serviceAccount);
+
+      LOG.info("[AUDIT] Workflow {} refers to secret {} of {}", workflowInstance.workflowId(),
+          secretName, serviceAccount);
+
+      final boolean createSecret;
+
+      // Check if we have a valid service account key secret already
+      final Secret secret = client.secrets().withName(secretName).get();
+      if (secret != null) {
+        if (!serviceAccountKeySecretIsValid(workflowInstance, serviceAccount, secretName, secret)) {
+          LOG.info("[AUDIT] Deleting invalid service account secret {} of {} for Workflow {}",
+              secretName, serviceAccount, workflowInstance.workflowId());
+          // Need to delete this secret before creating a new one
+          client.secrets().delete(secret);
+          createSecret = true;
+        } else {
+          createSecret = false;
+        }
+      } else {
+        createSecret = true;
+      }
+
+      // Create service account keys and secret if necessary
+      if (createSecret) {
+        createServiceAccountKeySecret(serviceAccount, secretName);
+        LOG.info("[AUDIT] Secret {} created for {} referred to by workflow {}",
+            secretName, serviceAccount, workflowInstance.workflowId());
+      }
+    });
+  }
+
+  private boolean serviceAccountKeySecretIsValid(WorkflowInstance workflowInstance, String serviceAccount,
+      String secretName, Secret secret) {
+    final Map<String, String> annotations = secret.getMetadata().getAnnotations();
+    final Map<String, String> data = secret.getData();
+    final String storedServiceAccount = annotations.get(STYX_WORKFLOW_SA_SECRET_ANNOTATION);
+    if (!storedServiceAccount.equals(serviceAccount)) {
+      LOG.error("[AUDIT] Workflow service account secret name annotation mismatch for workflow {} "
+              + "on secret {}, expected {} got {}",
+          workflowInstance.workflowId(), secretName, serviceAccount, storedServiceAccount);
+      return false;
     }
+
+    final String jsonKeyJson = data.get(STYX_WORKFLOW_SA_JSON_KEY);
+    final String p12KeyJson = data.get(STYX_WORKFLOW_SA_P12_KEY);
+    if (jsonKeyJson == null || p12KeyJson == null) {
+      return false;
+    }
+
+    final ServiceAccountKey jsonKey;
+    final ServiceAccountKey p12Key;
+    try {
+      jsonKey = Json.OBJECT_MAPPER.readValue(jsonKeyJson, ServiceAccountKey.class);
```

Comment:
It's only private key gets stored in kube secret, not the complete JSON. https://github.com/spotify/styx/pull/160/files#diff-d96c43ca98b2c4d50df1a802530ab077R266

Get key name from annotation instead?